skipchain/skipchain.go

```diff
@@ -381,6 +390,95 @@ func (s *Service) StoreSkipBlockInternal(psbd *StoreSkipBlock) (*StoreSkipBlockR
 	return reply, nil
 }
 
+// sendForwardLinkRequest sends a request to the given until until either the forward-link is
+// created or there's not enough online nodes to get a valid signature.
+func sendForwardLinkRequest(ro *onet.Roster, req *ForwardSignature, reply *ForwardSignatureReply) (err error) {
+	cl := NewClient()
+
+	// Try as many time as it can until the faulty threshold is reached
+	// meaning it's impossible to get a valid signature.
+	retries := protocol.DefaultFaultyThreshold(len(ro.List)) + 1
+
+	for i := 0; i < retries; i++ {
+		// No random permutation as we need a given threshold anyway.
+		err = cl.SendProtobuf(ro.List[i], req, reply)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+// OptimizeProof creates missing forward links to optimize the proof of the block
```

Comment:
I'm concerned about who is allowed to call this. It seems like anyone can call it, not just admins? Is that potentially expensive? Dangerous?

Reply:
That was also a concern for me but we have a handler opened for `ForwardSignature` requests anyway and this will only try to optimize what is necessary. Yes you could ask to optimize any block in the chain but eventually all forward links will be created and it won't start anything. In the worst case one or more can't be created, well an attacker could already send a `ForwardSignature` message to start the creation process. If that makes sense.